design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,224 @@
+means that the API that we design will not be able to leverage any "embedded-ness" of Connect,
+nullifying most of the benefits of the second approach. 
+
+There are still, however, benefits to packaging connect and KSQL in the same JVM for a seamless
+experience for new users. Namely, new users will only need to download a single package, spin up
+a single process and maintain only a single application. There are some (solvable) concerns with
+this approach, but for the reason outlined above, the syntax will be developed with the assumption
+that Connect is running as a separate component.
+
+**The recommendation is to support 1 & 3, tabling 2 unless a need for it comes in the future**
+
+### Syntax
+
+Analogous to DDL and DML, the Connect syntax will separate actions that have side effects with
+actions that simply declare metadata on top of existing data. This solves two issues:
+
+- A single connector can have many topics, which makes modeling them as a single statement difficult
+- A connector can be modeled in different ways (e.g. stream/table) and connector lifecycle should
+  be independent with how KSQL interprets it
+
+There are a few more minor points of consideration:
+- All "DML" commands that cause changes in connect will not be submitted to the command topic;
+  instead they will be issued on the REST server (this follows the pattern that all external-state
+  modifications are not distributed)
+- Some connectors will have KSQL-specific wrapper implementations that will allow simpler 
+  configuration and tighter integration. These connectors will also automatically issue the DDL 
+  statements for any topics that they create, when possible.
+
+**Sample Ingress**
+
+The first step is to create and start a Connector via the REST API (`POST /connectors`)
+```sql
+ksql> CREATE EXTERNAL SOURCE my-postgres-jdbc WITH ( \
+      connector.class="io.confluent.connect.jdbc.jdbcSourceConnector", \ 
+      tasks.max=1, \
+      connection.url=jdbc:postgresql://localhost:5432/my-db, \
+      ...);
+
+--------------
+SOURCE CREATED
+--------------
+```
+
+If the connector is one of the KSQL supported connectors, we can have default configuration options
+and tighter integration specified by `CREATE EXTERNAL SOURCE(type='JDBC') WITH (...);`.
+
+*Note: as Connect manages to simplify the configuration, whether through templates or any other
+means, we will immediately be able to leverage this simplification in KSQL's WITH clause.*
+
+Next, we provide tools to understand the status of the connector and corresponding topics hitting
+the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
+```sql
+ksql> SHOW EXTERNAL SOURCES;
+
+    name         |              connector                        |   topic prefix
+-----------------|-----------------------------------------------|--------------------
+my-postgres-jdbc | io.confluent.connect.jdbc.JdbcSourceConnector | test-postgres-jdbc
+
+ksql> DESCRIBE my-postgres-jdbc;
```

Comment:
Would this conflient with streams that have the same name? Let's say you have a `p1` connector for testing and a `p1` stream in KSQL. What would DESCRIBE describe?

Reply:
good point... I'm not sure how to address this issue. Suggestions for an alternative syntax?

Reply:
`DESCRIBE CONNECTOR <connector-name>`?

Snowflake has a reserved word after `DESCRIBE` for each object type they support. See https://docs.snowflake.net/manuals/sql-reference/sql/desc.html

I see other databases do not have anything like that. But, their objects are part of their servers, and they can provide object names checks to avoid conflicts with other objects types (such as table vs function names). However, in our case, we will have a Connector that is part of an external service, so KSQL cannot do any naming conflict checks. I think `DESCRIBE CONNECTOR <name>` would help.

Reply:
+1 for having a discriminator after `DESCRIBE`.  

Though if we're going with `CREATE EXTERNAL SOURCE` and `CREATE EXTERNAL SINK` then it should be `DESCRIBE EXTERNAL SOURCE foo` and `DESCRIBE EXTERNAL SINK bar`.   Which is a bit of a mouthful.

If it's `CREATE SOURCE CONNECTOR` we could get away with `DESCRIBE CONNECTOR`, assuming no name clash is possible between source and sink connectors.